File: ios/Classes/AMapParam.swift
//
//  AMapParam.swift
//  Runner
//
//  Created by xionghao on 2020/5/7.
//  Copyright © 2020 The Chromium Authors. All rights reserved.
//

class AMapParam: Codable {
    
    static let routeMap: Int = 0
    static let addressDescriptionMap: Int = 1
    
    let initialCenterPoint: Array<Double>
    let initialZoomLevel: CGFloat
    let enableMyLocation: Bool?
    let enableMyMarker: Bool?
    let mapType: Int
    let startAddressList: Array<AddressInfo>?
    let endAddressList: Array<AddressInfo>?
    let merchantAddressList: Array<AddressInfo>?
    let markerClickable: Bool
    
    init(
        initialCenterPoint: Array<Double>,
        initialZoomLevel: CGFloat,
        enableMyLocation: Bool? = false,
        enableMyMarker: Bool? = false,
        markerClickable: Bool = true,
        mapType: Int = routeMap,
        startAddressList: Array<AddressInfo>? = [],
        endAddressList: Array<AddressInfo>? = [],
        merchantAddressList: Array<AddressInfo>? = []
    ) {
        self.initialCenterPoint = initialCenterPoint
        self.initialZoomLevel = initialZoomLevel
        self.enableMyLocation = enableMyLocation
        self.enableMyMarker = enableMyMarker
        self.mapType = mapType
        self.startAddressList = startAddressList
        self.endAddressList = endAddressList
        self.merchantAddressList = merchantAddressList
        self.markerClickable = markerClickable
    }
}

class MarkerParam: Codable {
    let markerList: Array<AddressInfo>?
    
    init(markerList: Array<AddressInfo>? = []) {
        self.markerList = markerList
    }
}

struct AddressInfo: Codable, Hashable {
    static func == (lhs: AddressInfo, rhs: AddressInfo) -> Bool {
        return lhs.id == rhs.id
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    let geo: GeoPoint?
    let address: String?
    let index: Int?
    let indexName: String?
    let showType: Int?
    let id: Int?
    let parentId: Int?
    
}

struct GeoPoint: Codable, Hashable {
    let lat: Double?
    let lng: Double?
}

class AddressAnnotation {
    let address: AddressInfo
    let annotation: PointAnnotation
    
    init(address: AddressInfo, annotation: PointAnnotation) {
        self.address = address
        self.annotation = annotation
    }
}

File: lib/amap_param.dart

import 'package:flutter/cupertino.dart';

class AmapParam {
  static const int routeMap = 0;
  static const int addressDescriptionMap = 1;

  List<double> initialCenterPoint;
  double initialZoomLevel;
  bool enableMyLocation;
  bool enableMyMarker;
  int mapType;

  final List<AddressInfo> startAddressList;
  final List<AddressInfo> endAddressList;

  final List<AddressInfo> merchantAddressList;

  // final List<int> zoomShowTypes;

  AmapParam({
    @required this.initialCenterPoint,
    this.initialZoomLevel = 14,
    this.enableMyLocation = false,
    this.enableMyMarker = false,
    this.startAddressList = const [],
    this.endAddressList = const [],
    this.merchantAddressList = const [],
    this.mapType = routeMap
  });

  toJson() => <String, dynamic> {
    "initialCenterPoint": initialCenterPoint,
    "initialZoomLevel": initialZoomLevel,
    "enableMyLocation": enableMyLocation,
    "enableMyMarker": enableMyMarker,
    "mapType": mapType,
    "startAddressList": startAddressList,
    "endAddressList": endAddressList,
    "merchantAddressList": merchantAddressList
  };
}

class MarkerParam {
  final List<AddressInfo> markerList;
  MarkerParam(this.markerList);

  toJson() => <String, dynamic> {
    "markerList": markerList
  };
}

class AddressInfo {
  GeoPoint geo;
  String address;
  int index;
  String indexName;
  int showType;
  int id;
  int parentId;
  String provinceCode;
  AddressInfo(this.geo, this.address,{this.index = 0, this.indexName, this.showType, this.id, this.parentId, this.provinceCode});

  toJson() => <String, dynamic> {
    "geo": geo,
    "address": address,
    "index": index,
    "indexName": indexName,
    "showType": showType,
    "id": id,
    "parentId": parentId
  };
}

class GeoPoint {
  double lat;
  double lng;
  GeoPoint(this.lat, this.lng);

  toJson() => <String, dynamic> {
    "lat": lat,
    "lng": lng
  };
}